lua/ttt2/libraries/thermalvision.lua

```diff
@@ -0,0 +1,263 @@
+---
+-- This is the <code>thermalvision</code> library.
+-- It offers the possibility of rendering entities giving them the typical thermal vision look
+-- @author LeBroomer
+
+local render = render
+local table = table
+local IsValid = IsValid
+local cam = cam
+local surface = surface
+local hook = hook
+
+if SERVER then
+	AddCSLuaFile()
+else
+	thermalvision = {}
+
+	THERMALVISION_MODE_BOTH = 0
+	THERMALVISION_MODE_NOTVISIBLE = 1
+	THERMALVISION_MODE_VISIBLE = 2
+
+	local bgColoring = false
+	local thermalvisionList = {}
+	local thermalvisionHookInstalled = false
+
+	local overlayColorWithBG = Color(0, 0, 50, 235)
+	local overlayColorWithoutBG = Color(0, 0, 0, 235)
+
+	local colorModifyBG = {
+		["$pp_colour_addr"] = 0,
+		["$pp_colour_addg"] = 0.0,
+		["$pp_colour_addb"] = 0.8,
+		["$pp_colour_brightness"] = 0.1,
+		["$pp_colour_contrast"] = 0.15,
+		["$pp_colour_colour"] = 0.0,
+		["$pp_colour_mulr"] = 0,
+		["$pp_colour_mulg"] = 0,
+		["$pp_colour_mulb"] = 1
+	}
+
+	local colorModifyEnts = {
+		["$pp_colour_addr"] = 0.4,
+		["$pp_colour_addg"] = 0.2,
+		["$pp_colour_addb"] = 0.0,
+		["$pp_colour_brightness"] = 1.0,
+		["$pp_colour_contrast"] = 0.35,
+		["$pp_colour_colour"] = 30,
+		["$pp_colour_mulr"] = 0,
+		["$pp_colour_mulg"] = 0,
+		["$pp_colour_mulb"] = 0
+	}
+	---
+	-- Hook that renders the entities with the thermalvision
+	-- @realm client
+	local function RenderHook()
+		render.ClearStencil()
+		render.SetStencilEnable(true)
+		render.SetStencilWriteMask(255)
+		render.SetStencilTestMask(255)
+		render.SetStencilReferenceValue(1)
+
+		--first render pass: draw all entities even when z test fails
+		if #thermalvisionList > 0 then
+			--draw stencil mask and let everything through
+			render.SetStencilCompareFunction(STENCIL_ALWAYS)
+			render.SetStencilZFailOperation(STENCIL_REPLACE)
+			render.SetStencilPassOperation(STENCIL_REPLACE)
+			render.SetStencilFailOperation(STENCIL_ZERO)
+
+			--omit depth test but block writing to depth buffer
+			render.SuppressEngineLighting(true)
+			render.OverrideDepthEnable(true, false)
+			cam.Start3D()
+			cam.IgnoreZ(true)
+				for i = 1, #thermalvisionList do
+					local entry = thermalvisionList[i]
+					local ent = entry.ent
+					local mode = entry.mode
+
+					if not IsValid(ent) or mode == THERMALVISION_MODE_VISIBLE then continue end
```

Comment:
Theoretically the `ent`s in the list should be all the time valid...

Reply:
why? The entites could be removed after insertion into the list and hence could be invalid